penty/pentypes/builtins.py

```diff
@@ -472,21 +479,246 @@ def set_init(elts_ty=None):
+        raise TypeError
+    return self_ty
+
+def set_ixor(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not issubclass(other_ty, set):
+        raise TypeError
+    self_ty.__args__[0] |= other_ty.__args__[0]
+    return self_ty
+
+def set_len(self_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0]:
+        return _Cst[0]
+    return int
+
+def set_iter(self_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return {_SetIterator[ty] for ty in self_ty.__args__[0]}
+
+def set_or(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not issubclass(other_ty, set):
+        raise TypeError
+    return _Set[self_ty.__args__[0] | other_ty.__args__[0]]
+
+def set_sub(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not issubclass(other_ty, set):
+        raise TypeError
+    return _Set[self_ty.__args__[0]]
+
+def set_xor(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not issubclass(other_ty, set):
+        raise TypeError
+    return _Set[self_ty.__args__[0] | other_ty.__args__[0]]
+
+def set_add(self_ty, value_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    self_ty.__args__[0].add(_astype(value_ty))
+    return _Cst[None]
+
+def set_clear(self_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return _Cst[None]
+
+def set_copy(self_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return _Set[self_ty.__args__[0].copy()]
+
+def set_difference(self_ty, *other_tys):
+    from penty.penty import Types
+    if not issubclass(self_ty, set):
+        raise TypeError
+    for other_ty in other_tys:
+        if '__iter__' not in Types[other_ty]:
+            raise TypeError
+    return _Set[self_ty.__args__[0].copy()]
+
+def set_difference_update(self_ty, *other_tys):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    set_difference(self_ty, *other_tys)
+    return _Cst[None]
+
+def set_discard(self_ty, elt_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0]:
+        raise TypeError
+    return _Cst[None]
+
+def set_intersection(self_ty, *other_tys):
+    from penty.penty import Types
+    if not issubclass(self_ty, set):
+        raise TypeError
+    intersection_tys = self_ty.__args__[0].copy()
+    for other_ty in other_tys:
+        if '__iter__' not in Types[other_ty]:
+            raise TypeError
+        iter_tys = Types[other_ty]['__iter__'](other_ty)
+        if not isinstance(iter_tys, set):
+            iter_tys = {iter_tys}
+        for iter_ty in iter_tys:
+            if '__next__' not in Types[iter_ty]:
+                raise TypeError
+            value_ty = Types[iter_ty]['__next__'](iter_ty)
+            intersection_tys.add(value_ty)
```

Comment:
I guess a comment that explains why a union is done here instead of an intersection could make sense (it is quite counter-intuitive otherwise).